### PR TITLE
Fix CMakeLists.txt to not exclude everything with 'build' in pathname

### PR DESCRIPTION
**Pull request by leolchat**
The cmake globbing then excluding strategy implies that if the directory
in which we are compiling has a folder named `build` on its path like `/home/alice/project/build/folly/...`, everything breaks.
Note that it is the same for `tools` and `test`, I also fixed the tools one, but the test once has many files.... Using a file list would be the real solution.

### CODE REVIEW

**Comment by Orvid** [comment]
Where is this causing issues? The glob is specifically for building Folly's base library, and things in a `tools/` directory are binaries to run, and the build directory is the same way.

**Comment by leolchat** [comment]
This will fail with an empty list of files to compile:
```
cd /tmp
mkdir build
cd build
git clone https://github.com/facebook/folly.git
cd folly
cmake .
```

Compiling out of sources would fail too as long as the folly source folder path has a folder named build.

**Comment by leolchat** [comment]
So what is the decision on that? are you going to merge it? Or maybe you want to do a more complete fix of the issues related to globbing then excluding? Or you consider not a big enough issue?

**Comment by Orvid** [comment]
I'd missed your response because something's up with my github notifications :( 

I think the most correct solution here (assuming CMake allows it) would be to not use the absolute paths of the files while filtering.

**Comment by leolchat** [comment]
Maybe you could do match `${CMAKE_CURRENT_LIST_DIR}/build` or something like that, not sure.

**Comment by leolchat** [comment]
your implementation in the end (the function `REMOVE_MATCHES_FROM_LISTS `) uses `if(<path> MATCHES test)` but MATCHES can take a regexp so `${CMAKE_CURRENT_LIST_DIR}/*/test/` would probably work.

**Comment by leolchat** [comment]
Should be bulletproof now.

File: CMakeLists.txt [hunk 1]
```diff
@@ -90,14 +90,14 @@ auto_sources(hfiles "*.h" "RECURSE" "${FOLLY_DIR}")
...
-    "/futures/exercises/"
-    "/test/"
-    "/tools/"
+    "^${FOLLY_DIR}/build/"
+    "^${FOLLY_DIR}/experimental/exception_tracer/"
+    "^${FOLLY_DIR}/experimental/hazptr/bench/"
+    "^${FOLLY_DIR}/experimental/hazptr/example/"
+    "^${FOLLY_DIR}/experimental/logging/example/"
+    "^${FOLLY_DIR}/futures/exercises/"
+    "^${FOLLY_DIR}.*/test/"
```
Comment (yfeldblum) [inline CMakeLists.txt]: Should this be `}/.*/`?

Comment (leolchat) [reply CMakeLists.txt]: no because there is `${FOLLY_DIR}/test/` to match

Comment (yfeldblum) [reply CMakeLists.txt]: Gotcha. Consider `}/(.*/)?`?

**Pull request closed without merging**